libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -204,31 +207,56 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
         jobControl.scope.launch(Dispatchers.Main) {
             mutex.withLock {
                 update.onValue {
-                    updateLineGradient(
-                        style,
-                        RouteLayerConstants.PRIMARY_ROUTE_TRAFFIC_LAYER_ID,
+                    getExpressionUpdateFun(
                         it.primaryRouteLineDynamicData.trafficExpressionProvider
-                    )
-                    updateLineGradient(
-                        style,
-                        RouteLayerConstants.PRIMARY_ROUTE_LAYER_ID,
+                    ).apply {
+                        this(
+                            style,
+                            RouteLayerConstants.PRIMARY_ROUTE_TRAFFIC_LAYER_ID,
+                            it.primaryRouteLineDynamicData.trafficExpressionProvider
+                        )
+                    }
+                    getExpressionUpdateFun(
                         it.primaryRouteLineDynamicData.baseExpressionProvider
-                    )
-                    updateLineGradient(
-                        style,
-                        RouteLayerConstants.PRIMARY_ROUTE_CASING_LAYER_ID,
+                    ).apply {
+                        this(
+                            style,
+                            RouteLayerConstants.PRIMARY_ROUTE_LAYER_ID,
+                            it.primaryRouteLineDynamicData.baseExpressionProvider
+                        )
+                    }
+                    getExpressionUpdateFun(
                         it.primaryRouteLineDynamicData.casingExpressionProvider
-                    )
-                    updateLineGradient(
-                        style,
-                        RouteLayerConstants.RESTRICTED_ROAD_LAYER_ID,
+                    ).apply {
+                        this(
+                            style,
+                            RouteLayerConstants.PRIMARY_ROUTE_CASING_LAYER_ID,
+                            it.primaryRouteLineDynamicData.casingExpressionProvider
+                        )
+                    }
+                    getExpressionUpdateFun(
                         it.primaryRouteLineDynamicData.restrictedSectionExpressionProvider
-                    )
+                    ).apply {
+                        this(
+                            style,
+                            RouteLayerConstants.RESTRICTED_ROAD_LAYER_ID,
+                            it.primaryRouteLineDynamicData.restrictedSectionExpressionProvider
+                        )
+                    }
                 }
             }
         }
     }
 
+    private fun getExpressionUpdateFun(
+        provider: RouteLineExpressionProvider?
+    ): (Style, String, RouteLineExpressionProvider?) -> Unit {
+        return when (provider) {
+            is RouteLineTrimExpressionProvider -> ::updateTrimOffset
+            else -> ::updateLineGradient
```

Comment:
Will the logic in this class still work if a regular gradient expression was provided in `RouteLineUpdateValue`? Would the layers be displayed correctly or there's a risk of trail layer messing styling up? Since `RouteLineUpdateValue` is public and mutable, developers could provide their own gradient expression that would override our new `RouteLineTrimExpressionProvider` expression.

Reply:
The `Expression` syntax for the trim expression is very different than the expression syntax that has been used thus far.  Also the maps API call is different. Both expression types are supported but  the `RouteLineTrimExpressionProvider` is expected to provide a trim type expression and the original `RouteLineExpressionProvider` is expected to produce the original type.  

When displaying the traffic congestion the "regular" type of expression is still used.  Also when using the feature to style inactive route legs differently than the active route leg the regular type of expression is produced.  I tested with the QA app. which has examples for active/inactive route leg styling as well as route restrictions and all of the example work normally.

Here's a complex example from the QA app.  The inactive leg is styled differently.  The magenta is a restricted section. The orange and blue are the traffic congestion. The grey behind the puck are the colors for the "vanished" part of the route line. In this case the trail layers introduced haven't caused any problems.

![Screenshot_20220504-100600](https://user-images.githubusercontent.com/2249818/166653176-3e5b72f1-b7fd-4af9-9fd7-1f17fb49299a.png)


> is public and mutable, developers could provide their own gradient expression

This has always been true so a developer could produce an incorrect `Expression`. The introduction of the trim offset feature complicates this a bit.  The trim offset expression is easier to create but it certainly is possible a developer could create the wrong type of expression for the provider used. I don't see a way around this, at least not without breaking semver. The maps API implementation requires different syntax and different API calls.  While I tried to abstract the complexity, I couldn't think of anything better. Fundamentally nothing has really changed though. If a developer is going to override the expression produced by the API they must produce a valid `Expression`. This detail hasn't changed.